lib/resty/wasmx/shm.lua

```diff
@@ -0,0 +1,471 @@
+    if type(name) ~= "string" or name == "" then
+        error("name must be a non-empty string", 2)
+    end
+
+    if metric_type_set[metric_type] == nil then
+        error("metric_type must be either \
+               resty.wasmx.shm.metrics.COUNTER, \
+               resty.wasmx.shm.metrics.GAUGE or \
+               resty.wasmx.shm.metrics.HISTOGRAM", 2)
+    end
+
+    local cname = ffi_new("ngx_str_t", { data = name, len = #name })
+
+    local m_id = C.ngx_wa_ffi_shm_define_metric(cname, metric_type)
+    if m_id == FFI_ERROR then
+        return nil, "failed defining metric"
+    end
+
+    return m_id
+end
+
+
+local function record_metric(zone, metric_id, value)
+    if type(metric_id) ~= "number" then
+        error("metric_id must be a number", 2)
+    end
+
+    if type(value) ~= "number" then
+        error("value must be a number", 2)
+    end
+
+
+    local rc = C.ngx_wa_ffi_shm_record_metric(metric_id, value)
+    if rc == FFI_ERROR then
+        return nil, "failed recording metric"
+    end
+
+    if rc == FFI_DECLINED then
+        return nil, "metric not found"
+    end
+
+    return true
+end
+
+
+local function increment_metric(zone, metric_id, value)
+    local increment = 1
+
+    if type(metric_id) ~= "number" then
+        error("metric_id must be a number", 2)
+    end
+
+    if value ~= nil then
+        if type(value) ~= "number" or value < 1 then
+            error("value must be greater than zero", 2)
+        end
+
+        increment = value
+    end
+
+
+    local rc = C.ngx_wa_ffi_shm_increment_metric(metric_id, increment)
+    if rc == FFI_ERROR then
+        return nil, "failed incrementing metric"
+    end
+
+    if rc == FFI_DECLINED then
+        return nil, "metric not found"
+    end
+
+    return true
+end
+
+
+local get_zones_handler = ffi_cast("ngx_wa_ffi_shm_get_zones_handler_pt",
+function(shm)
+    local zone_name = ffi_str(shm.name.data, shm.name.len)
+    _M[zone_name] = { [WASM_SHM_KEY] = shm }
+
+    if shm.type == types.ffi_shm.SHM_TYPE_KV then
+        _M[zone_name].get_keys = get_keys
+        _M[zone_name].get = get_kv_value
+        _M[zone_name].set = set_kv_value
+
+    elseif shm.type == types.ffi_shm.SHM_TYPE_QUEUE then
+        -- NYI
+
+    elseif shm.type == types.ffi_shm.SHM_TYPE_METRICS then
+        _M[zone_name].define = define_metric
+        _M[zone_name].record = record_metric
+        _M[zone_name].increment = increment_metric
+        _M[zone_name].get_keys = get_keys
+        _M[zone_name].get = get_metric
+        _M[zone_name].get_by_name = get_metric_by_name
+        _M[zone_name].COUNTER = types.ffi_metric.COUNTER
+        _M[zone_name].GAUGE = types.ffi_metric.GAUGE
+        _M[zone_name].HISTOGRAM = types.ffi_metric.HISTOGRAM
+    end
+end)
+
+
+function _M.init()
+    if initialized then
+        return
+    end
+
+    local rc = C.ngx_wa_ffi_shm_get_zones(get_zones_handler)
+    if rc == FFI_DECLINED then
+        ngx_log(ngx.INFO, "no shm zones found")
```

Comment:
Maybe a debug log at most? This log will also not have any context so we might want to add some, something like "no shm zones found for resty.wasmx interface"